### ISSUE TITLE
Metastore Manager and Hive Editor assist do not display DB tables for non-default databases unless ALL DB Ranger Hive policy access 

### ISSUE THREAD

**Issue by elisabetao** [cloudera/hue#883]
Hi,

Metastore Manager and Hive Editor assist do not display DB tables for non-default databases unless Ranger Hive policy is set to ALL DB access. If access is granted to a single or multiple DBs (but not ALL - *) Hue does not display the tables list, however running: `use DB; show tables;` in the Hive Query Editor and we are able to properly have the tables in the results. Running queries against the DBs and tables works fine, as well.
We enabled debug but there's no additional info that could help trace the problem.

Thanks

**Comment by romainr** [comment]
@weixiaxu would be great to confirm if the HS2 GetSchemas() call used by Hue as an issue with Ranger

https://github.com/cloudera/hue/blob/master/apps/beeswax/src/beeswax/server/hive_server2_lib.py#L757 and get_tables

**Comment by elisabetao** [comment]
Hi Romain and Wei,

Do you have some hints on how we can test/check further?

Thanks a lot

**Comment by romainr** [comment]
This is similar to https://issues.cloudera.org/browse/HUE-6256, where some corner cases might still be there

**Comment by weixiaxu** [comment]
Hi,

Can you please share a bit more about your env? I tried with new user t1 belongs to grp1, and grant all permission for 't1' user with 'testdb' access only, not 'all'. Log in as t1, we can do any operations as granted/expected for 'testdb' only. Make sure this user/group exists on your cluster OS level.

Hope this helps!

Weixia

**Comment by elisabetao** [comment]
Hi Weixia and Romain,

Thanks a lot for your quick replies. We actually have SSSD-AD OS integrations, as such the users and groups membership already exists . We can also see the Hive DBs list and can perform  operations on the DBs and tables, however we are no able to see the tables and their fields displayed under the corresponding   Hive DBS Hue left pane or the Table Browser (no entries found), for users that do not have ALL DBs permissions.  Nevertheless, we are able to properly consult the default DB tables and fields in the above mentioned Hive DBs pane and in the Table Browser.  Is there any specific configuration that would allow a different behaviour from Hue for the default DB? 
![test](https://user-images.githubusercontent.com/24717881/60213055-ab38f680-9830-11e9-8898-cac9b9fffc41.png)

Thanks a lot

**Comment by weixiaxu** [comment]
Hi Elisabetao,

Are you using a secure HDP cluster(kerberos +ranger) with Hue?  Which version? Ranger integration is not fully certified. We could have an issue on hive assist panel.

Cheers,

Weixia

**Comment by elisabetao** [comment]
Hi Weixia,

Indeed, we are on HDP 3.1.1 secure ( Kerberos + ranger) with Hue 4.3. Would you mind giving more details about this issue about the assist panel? 
Is there a workaround/patch code we could apply?

Thanks so much,
Elisabeta

**Comment by weixiaxu** [comment]
Hi Elisabeta,

Unfortunately HDP+Hue is not a supported combination as of now.  We do have an internal JIRA filed  against hive/hue ranger integration which has not being worked on yet.  We will keep you update once it gets resolved.

Sorry for the inconvenience!

Weixia

**Comment by elisabetao** [comment]
Hi Weixia,

Thanks to you for the quick reply and help. Since we are already preparing this for our Prod deployment could you maybe help pointing me to which are the assist panel functions/code we could look into to correct and why this is working propely for the Default Hive DB ? 

Thanks  a lot,
Elisabeta

**Comment by elisabetao** [comment]
Hi Weixia,

Do you happen to have visibility to when this issue will be fixed? The fact that we are not able to properly view the Hive tables in assist while having access to them can be quite frustrating when accessing Hive through Hue.

Thanks a lot for the help,
Elisabeta

**Comment by weixiaxu** [comment]
Hi Elisabeta,

HDP with HUE is not on our plan as of now. We tried with our new product, after config hive-site properly we are able to see granted db/tables(by Ranger) for the specific user. So please check if you have "hive.server2.authentication.kerberos.principal" setup properly.   If issue still exists, you can check apps/beeswax/src/beeswax/server/hive_server2_lib.py or related files to understand more.

Hope this helps.

Weixia


Hope this helps!

Weixia

**Comment by leopaul36** [comment]
Hi @romainr @weixiaxu, we also encountered this issue with the latest version of Hue and HDP 3.1. 

How would this work with CDP which has both Hue and Ranger in the stack ?

Thanks for your help.

**Comment by elisabetao** [comment]
Hi,

To overcome this, I have replaced the use of thrift GetTables based call from get_tables_meta in `https://github.com/cloudera/hue/blob/master/apps/beeswax/src/beeswax/server/dbms.py`  with :
```
def get_tables_meta(self, database='default', table_names='*', table_types=None):
    database = database.lower() # Impala is case sensitive

    if self.server_name == 'beeswax':
      identifier = self.to_matching_wildcard(table_names)
    else:
      identifier = None
    #tables = self.client.get_tables_meta(database, identifier, table_types)
    #Switch back to show tables/views since GetTables does not return all tables/views in DB for non-default databases unless ALL DB Ranger Hive policy access 
    query_tbls = "show tables IN "+str(database)
    query = hql_query(query_tbls)
    timeout = SERVER_CONN_TIMEOUT.get()
    handle_tbls = self.execute_and_wait(query, timeout_sec=timeout)
    if handle_tbls:
      result = self.fetch(handle_tbls, rows=5000)
      self.close(handle_tbls)
      tbls = [[table_name[0],'Table'] for table_name in result.rows()]

    query_vws = "show views IN "+str(database)
    query = hql_query(query_vws)
    handle_vws = self.execute_and_wait(query, timeout_sec=timeout)
    if handle_vws:
      result = self.fetch(handle_vws, rows=5000)
      self.close(handle_vws)
      vws = [[view_name[0],'View'] for view_name in result.rows()]
    
    query_mvws = "show materialized views IN "+str(database)
    query = hql_query(query_mvws)
    handle_mvws = self.execute_and_wait(query, timeout_sec=timeout)
    if handle_mvws:
        result = self.fetch(handle_mvws, rows=5000)
        self.close(handle_mvws)
        mvws = [[mview_name[0],'Materialized view'] for mview_name in result.rows() if str(mview_name[0])  and  str(mview_name[0]) != "# MV Name" ]

    tables = tbls + vws + mvws
    hive_objects = [{'comment': None, 'type':table[1], 'name':table[0]} for table in tables]
    

    if len(hive_objects) <= APPLY_NATURAL_SORT_MAX.get():
      hive_objects = apply_natural_sort(hive_objects, key='name')

    #if len(tables) <= APPLY_NATURAL_SORT_MAX.get():
    #  tables = apply_natural_sort(tables, key='name')
    #return tables
    return hive_objects
```

Hope this helps.

Thanks

**Comment by NLDJAI1** [comment]
Hi Elisabetao,

Thanks a lot Elisabetao. The workaround indeed resolves the issue after pressing the refresh icon on Tables at least once. But this is a bug and needs to be fixed by Cloudera as DAS LTE 1.3.0 version with HDP 3.1.0 is very immature and a lot of bugs. For the customers who cannot upgrade either HDP or DAS due to cascading effects to other software versions are getting this problem and need to look back to HUE as a workaround UI to support analysts.

**Comment by NLDJAI1** [comment]
Please proceed with closing the case.

Regards,
Deepak
